File: fixtures/python_demo/analysis/report.py
"""Stage 2a — a DOWNSTREAM consumer of processed_iris.csv.

Never imports the pipeline. Its only connection to it is the data product.
"""
from __future__ import annotations

from pathlib import Path

import pandas as pd

PROCESSED = Path(__file__).resolve().parents[1] / "processed_iris.csv"


def load_processed(path=PROCESSED):
    return pd.read_csv(path)


def petal_report(df=None):
    """Reads the petal_ratio and species columns."""
    df = load_processed() if df is None else df
    ranked = df.sort_values("petal_ratio", ascending=False)
    return ranked[["species", "petal_ratio", "sepal_ratio"]]


def widest_sepals(df=None):
    df = load_processed() if df is None else df
    return df.nlargest(5, "sepal_width")[["species", "sepal_width"]]

File: fixtures/r_demo/analysis/report.R
# Stage 2a — a DOWNSTREAM consumer of processed_iris.csv.
# Never sources the pipeline. Its only connection to it is the data product.

suppressPackageStartupMessages({
  library(dplyr)
  library(readr)
})

PROCESSED <- "processed_iris.csv"

load_processed <- function(path = PROCESSED) {
  read_csv(path, show_col_types = FALSE)
}

#' Reads the petal_ratio and species columns.
petal_report <- function(df = NULL) {
  if (is.null(df)) df <- load_processed()
  df %>%
    arrange(desc(petal_ratio)) %>%
    select(species, petal_ratio, sepal_ratio)
}

widest_sepals <- function(df = NULL) {
  if (is.null(df)) df <- load_processed()
  df %>%
    slice_max(sepal_width, n = 5) %>%
    select(species, sepal_width)
}
